migration/idempotent/017-metric-rollup.sql

```diff
@@ -0,0 +1,282 @@
+            )', r.schema_name) -- Get metric names that have a pending rollup creation.
+        LOOP
+            SELECT INTO rollup_view_created _prom_catalog.create_metric_rollup_view(r.schema_name, m.metric_name, m.table_name, r.resolution);
+            IF rollup_view_created THEN
+                EXECUTE FORMAT('INSERT INTO _prom_catalog.metric_with_rollup VALUES (%L, %L, %L, TRUE)', r.schema_name, m.metric_name, m.table_name);
+                new_rollups_created := new_rollups_created + 1;
+            END IF;
+        END LOOP;
+        RAISE LOG '[Rollup] Created % for rollup name %', new_rollups_created, r.name;
+    END LOOP;
+
+    COMMIT;
+
+    -- Refresh the newly added rollups for entire duration of data.
+    FOR r IN
+        SELECT rollup_schema, table_name FROM _prom_catalog.metric_with_rollup WHERE refresh_pending = TRUE
+    LOOP
+        CALL refresh_continuous_aggregate(r.rollup_schema || '.' || r.table_name, NULL, NULL);
+        UPDATE _prom_catalog.metric_with_rollup SET refresh_pending = FALSE WHERE rollup_schema = r.rollup_schema AND table_name = r.table_name;
+    END LOOP;
+END;
+$$
+LANGUAGE PLPGSQL;
+
+CREATE OR REPLACE FUNCTION prom_api.set_automatic_downsample(state BOOLEAN)
+RETURNS BOOLEAN
+VOLATILE
+SET search_path = pg_catalog, pg_temp
+AS $$
+    SELECT _prom_catalog.set_default_value('automatic_downsample', state::text);
+    SELECT true;
+$$
+LANGUAGE SQL;
+COMMENT ON FUNCTION prom_api.set_automatic_downsample(BOOLEAN)
+    IS 'Set automatic downsample state for metrics (a.k.a. metric rollups). Metric rollups will be created only if this is true';
+GRANT EXECUTE ON FUNCTION prom_api.set_automatic_downsample(BOOLEAN) TO prom_admin;
+
+CREATE OR REPLACE FUNCTION prom_api.get_automatic_downsample()
+RETURNS BOOLEAN
+SET search_path = pg_catalog, pg_temp
+AS $func$
+    SELECT _prom_catalog.get_default_value('automatic_downsample')::boolean;
+$func$
+LANGUAGE SQL;
+COMMENT ON FUNCTION prom_api.get_automatic_downsample()
+    IS 'Get automatic downsample state for metrics (a.k.a. metric rollups)';
+GRANT EXECUTE ON FUNCTION prom_api.get_automatic_downsample() TO prom_admin;
+
+CREATE OR REPLACE FUNCTION prom_api.set_downsample_resolution(resolutions TEXT)
+    RETURNS BOOLEAN
+    VOLATILE
+    SET search_path = pg_catalog, pg_temp
+AS $$
+    SELECT _prom_catalog.set_default_value('downsample_resolution', resolutions::text);
+SELECT true;
+$$
+LANGUAGE SQL;
+COMMENT ON FUNCTION prom_api.set_downsample_resolution(TEXT)
+    IS 'Set downsample resolution for metric rollups. This must be a comma separated string of label:resolution:retention';
+GRANT EXECUTE ON FUNCTION prom_api.set_downsample_resolution(TEXT) TO prom_admin;
+
+CREATE OR REPLACE FUNCTION prom_api.get_downsample_resolution()
+    RETURNS TEXT
+    SET search_path = pg_catalog, pg_temp
+AS $func$
+    SELECT _prom_catalog.get_default_value('downsample_resolution')::TEXT;
```

Comment:
I don't see an initial default for this. Is that intended?

Reply:
Yes, this is intended, but I am not confident that this is the correct pattern.

The idea behind this is, `downsample_resolution` (example value `short:5m:90d,long:1h:365d`) will be set by connector (by seeing the `dataset-config`). We need to store this somewhere in the DB, so that when the user changes the downsample (in `dataset-config`) to something like `short:5m:90d,very_long:1d:365d`, then connector deletes the `long` based downsample and creates the `very_long` downsample resolution. Hence, we need a place to store the current resolution state the DB contains.

There should not be any entry in `_prom_catalog.initial_defaults` view, since when Promscale starts, we do not have any downsampling. Creation of downsampling is done by connector.

Imagine, if we had a initial default for downsample_resolution, then Promscale starts and tries to delete the downsampling if the dataset-config is saying a different resolution than a default, and that creates an exception in the database.

I didn't wanna create a new table just to keep the state, hence used the defaults table.

Reply:
> ach downsample resolution could be a separate row instead of using a comma-separated string

Ah, now I relate that i am already creating a `_prom_catalog.rollup` table that can be used for doing exactly this. It has all the information that the downsample resolution string needs i.e., label, resolution, retention. So, we should not store any of this in defaults table.

Thank you for the pointer.